dm/worker/server.go

```diff
@@ -98,21 +98,17 @@ func (s *Server) Start() error {
 		// worker keepalive with master
 		// If worker loses connect from master, it would stop all task and try to connect master again.
 		shouldExit := false
+		var err1 error
 		for !shouldExit {
-			shouldExit, err = s.KeepAlive()
-			if err != nil || !shouldExit {
+			shouldExit, err1 = s.KeepAlive()
+			if err1 != nil || !shouldExit {
 				if s.retryConnectMaster.Get() {
 					// Try to connect master again before stop worker
 					s.retryConnectMaster.Set(false)
 				} else {
-					var w *Worker
-					s.Lock()
-					if s.worker != nil {
-						w = s.worker
-						s.worker = nil
-					}
-					s.Unlock()
+					w := s.getWorker(true)
 					if w != nil {
+						s.setWorker(nil, true)
```

Comment:
Is it possible to overwrite another `w` before this `setWorker` but after `getWorker` in L109?

Reply:
The logic has been revised in https://github.com/pingcap/dm/pull/453/commits/5335b65328f3fcf25b713dcd1498d6ab9cbf4e6b